### ISSUE TITLE
Problems with Test on Firefox 82

### ISSUE THREAD

**Issue by cproof** [rtr-nettest/rmbtws#15]
Multiple users report the websocket-test failing on Firefox 82.

Steps to reproduce:
* Start the test on Firefox 82
* Test will fail during init_up or up phase

The same test worked with Firefox 81.
In the [release notes](https://www.mozilla.org/en-US/firefox/82.0/releasenotes/) released by Mozilla, no changes that should affect the test are listed:

**Comment by cproof** [comment]
The problem was identified and seems to be a bug in Firefox 82:

The rmbt protocol uses the last bit of a `chunk` to signal to the test server that a test phase should be complete (see [doc](https://www.netztest.at/doc/#termination_byte)). During the upload pre-test phase, the rmbtws client is sending chunk sizes for a fixed time, in the following order

* 1x 4KB
* 2x 4KB
* 4x 4KB
* 8x 4KB
* 8x 8KB
* 8x 16KB
* 8x 32KB
* 8x 64KB
* 8x 128KB
etc.

For the first `n-1` chunks of each phase (so 7 chunks for the most phases), the last bit of every chunk is set to `0x00` and for the last chunk of every phase, the last bit of the chunk is set to `0xff`. This is used to signal the [test-server](https://github.com/rtr-nettest/rmbt-server/) that one phase is finished, the `time` should be sent and the server loop resets for the new command.

Up until Firefox 81, this worked fine. However, with Firefox 82, after in-depth debugging, Wireshark shows that this works for up until 8x64KB, but starting with the 8x128KB phase, the last bit of the last chunk remains set to `0x00`.

I then built a [special version](https://github.com/rtr-nettest/rmbtws/tree/firefox82-reproduce) of the code, where all chunks are numbered in sequence just before being sent:

```javascript
for (let i = 0; i < total; i++) {
    let blob;
    if (i === (total - 1)) {

        let x = new Uint8Array(_arrayBuffers[chunkSize][0].slice(0));
        x[x.length-1] = 0xff

        blob = x.buffer;
    } else {
        blob = _arrayBuffers[chunkSize][0];
    }

    let y = new Uint8Array(blob)
    y[0] = i
    y[1] = thread.id
    console.log("bytes 0 1 -3 -2 -1: " + thread.id + " / " + i, y[0], y[1], y[y.length - 3], y[y.length - 2], y[y.length - 1])
    socket.send(y.buffer);
}
```

It can be run as follows
```bash
git clone https://github.com/rtr-nettest/rmbtws/
git checkout -b firefox82-reproduce origin/firefox82-reproduce
npm install
gulp
http-server
# now, open http://localhost:8080/test/Websockettest.html in Firefox 82, and click the "start" button
```

The debug output is as expected:

```
0: uploading 8 chunks, 524.288 KB rmbtws.js:955:17 //64KB chunks
bytes 0 1 -3 -2 -1: 0 / 0 0 0 32 190 0 rmbtws.js:942:29
bytes 0 1 -3 -2 -1: 0 / 1 1 0 32 190 0 rmbtws.js:942:29
bytes 0 1 -3 -2 -1: 0 / 2 2 0 32 190 0 rmbtws.js:942:29
bytes 0 1 -3 -2 -1: 0 / 3 3 0 32 190 0 rmbtws.js:942:29
bytes 0 1 -3 -2 -1: 0 / 4 4 0 32 190 0 rmbtws.js:942:29
bytes 0 1 -3 -2 -1: 0 / 5 5 0 32 190 0 rmbtws.js:942:29
bytes 0 1 -3 -2 -1: 0 / 6 6 0 32 190 0 rmbtws.js:942:29
bytes 0 1 -3 -2 -1: 0 / 7 7 0 32 190 255 rmbtws.js:942:29
0: TIME 225985631

0: uploading 8 chunks, 1048.576 KB rmbtws.js:955:17 //128KB chunks
bytes 0 1 -3 -2 -1: 0 / 0 0 0 130 140 0 rmbtws.js:942:29
bytes 0 1 -3 -2 -1: 0 / 1 1 0 130 140 0 rmbtws.js:942:29
bytes 0 1 -3 -2 -1: 0 / 2 2 0 130 140 0 rmbtws.js:942:29
bytes 0 1 -3 -2 -1: 0 / 3 3 0 130 140 0 rmbtws.js:942:29
bytes 0 1 -3 -2 -1: 0 / 4 4 0 130 140 0 rmbtws.js:942:29
bytes 0 1 -3 -2 -1: 0 / 5 5 0 130 140 0 rmbtws.js:942:29
bytes 0 1 -3 -2 -1: 0 / 6 6 0 130 140 0 rmbtws.js:942:29
bytes 0 1 -3 -2 -1: 0 / 7 7 0 130 140 255 rmbtws.js:942:29
```
 
also the Firefox F12 output seems fine:
![grafik](https://user-images.githubusercontent.com/4948782/97308362-1483ac00-1861-11eb-984d-562d3dede4ae.png)

But what is very strange is that when I look at the transfer with Wireshark, Firefox seems to have transmitted only chunks 1-4 of the 128KB chunks, and then repeatetly transmitted chunk 4 of the 128KB chunks:

![grafik](https://user-images.githubusercontent.com/4948782/97309366-3893bd00-1862-11eb-9edc-5c0f0478d1ba.png)

[Wireshark dump of Firefox 81](https://thomas.sinra.at/tmp/dump_fx81.pcapng)  
[Wireshark dump of Firefox 82](https://thomas.sinra.at/tmp/dump_fx82.pcapng)

Therefore, the server never receives the `0xff` bit and therefore is unable to exit the phase. This seems to occurr both beginning with 8x128KB chunks as well as with 16*64KB chunks. I'll still need to open a issue with Mozilla.

**Comment by cproof** [comment]
Bug openend at Mozilla here: https://bugzilla.mozilla.org/show_bug.cgi?id=1673621

**Comment by cproof** [comment]
"Bugfix" version available here: https://github.com/rtr-nettest/rmbtws/tree/firefox82-fix (here, for Firefox 82 #chunks is fixed to `8` and max chunk size is fixed to `64KB`).
As this may lead to inaccuracies with faster connections, if the upload pre-test calculates a UL speed lower than a quarter of the DL speed, the #threads and chunk size for the upload phase is calculated on the estimate of dl_speed/4 for upload.

**Comment by cproof** [comment]
Mozilla is aware of the bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1673340
The bug was set to `BLOCKING`

**Comment by cproof** [comment]
[Fixed in Firefox 82.0.2](https://www.mozilla.org/en-US/firefox/82.0.2/releasenotes/)

**Comment by helb** [comment]
Thanks, Thomas! Just tried FF 82.0.2 on netmetr.cz and indeed it works again :+1:
